### DISCUSSION TITLE
stats fmriprep aroma

### DISCUSSION THREAD

**Discussion by Rageve** [cpp-lln-lab/bidspm#991] [General]
Dear Rami,

now I'm starting to run the statistics using my own dataset. 
We conducted a food-cue reactivity task on two sessions and with two runs each. Preprocessing with fmriprep worked fine. We selected the flag --use-aroma to nonaggressively  correct movement related noise components. In this case fmriprep creates a file sub-1085_ses-V0_task-food_run-1_space-MNI152NLin6Asym_desc-smoothAROMAnonaggr_bold.nii that is already smoothed with FWHM of 6mm.
The preprocessed data is now in the folder

```
.../BIDS/Tuebingen/inputs/fmriprep/sub-1085/ses-V0/func/
.../BIDS/Tuebingen/inputs/fmriprep/sub-1085/ses-V9/func
```

In your default setup you expect smoothing, because fmriprep is normally not smoothing the data during preprocessing. To become more familiar with the stats processing I  created a simple default model 

```matlab
task_in_the_model ={'food'}

bidspm(bids_dir, output_dir, 'dataset', ...
        'action', 'default_model', ...
        'task', task_in_the_model, ...
        'space', {'MNI152NLin6Asym'});
```

the model was then created and saved at /server/fo2-17/data/NASAL_2018/DAGen_Bromo/BIDS/Tuebingen/outputs/derivatives/models/model-defaultFood_smdl.json
It's only a test file and there are some modifications in the design matrix (i.e. confounds) and contrast settings necessary (I use visual studio code for editing). 

then I tried to run the stats using

```matlab
bidspm(bids_dir, output_dir, 'subject', ...
         'participant_label', {subject_label}, ...
         'action', 'stats', ...
         'preproc_dir', prepro_dir, ...
         'task', {'food'}, ...
         'model_file', model_file);
```

and the script stopped with the error message

```
Indexing subject: sub-1085 [....]

[10:22:13] bidspm - INFO				getData
WILL WORK ON SUBJECTS
	- 1085


[10:22:13] bidspm - INFO				printWorkflowName
WORKFLOW: SUBJECT LEVEL GLM

[10:22:13] bidspm - INFO				bidsFFX

 PROCESSING NODE: run

Warning:
[10:22:13] bidspm - WARNING				bidsFFX
No data found in /server/fo2-17/data/NASAL_2018/DAGen_Bromo/BIDS/Tuebingen/inputs/fmriprep
for subject 1085 for filter:

	- desc: {smth6}
	- extension: {.nii.*}
	- modality: {func}
	- prefix: {}
	- space: {MNI152NLin6Asym}
	- sub: {^1085$}
	- suffix: {bold}
	- task: {food}

Available spaces are:

	- MNI152NLin2009cAsym
	- MNI152NLin6Asym

Available tasks are:
```

and ...

Now there are several question. Should I copy the data to a folder outside fmriprep (bidspm-preproc) and how? How can I change the input of
the - desc flag {smth6} to {smooth}. How can I adapt the script that only one session (ses-V0) will be analyzed?

Best
Ralf
[model-defaultFood_smdl.zip](https://github.com/cpp-lln-lab/bidspm/files/11342339/model-defaultFood_smdl.zip)

**Comment by Remi-Gau** [comment]
OK if you want to use fmriprep data without extra smoothing, you must first indeed copy the data you need in a bidspm-preproc folder.

For this you can use the copy action

https://bidspm.readthedocs.io/en/latest/usage_notes.html


```
task_in_the_model ={'food'}

bidspm(bids_dir, output_dir, 'subject', ...
        'action', 'copy', ...
        'participant_label', {subject_label}, ...
        'bids_filter_file', struct([]), ...
        'verbosity', 2, ...
        'space', {'MNI152NLin6Asym'}, ...
        'options', struct([]), ...
        'task', task_in_the_model, ...
        'skip_validation', true)
```

You can use the `'bids_filter_file', struct([])` to specify more filter for what data you want to copy.

The default structure used is this one:

```matlab
struct('fmap', struct('modality', 'fmap'), ...
       'bold', struct('modality', 'func', 'suffix', 'bold'), ...
       't2w',  struct('modality', 'anat', 'suffix', 'T2w'), ...
       't1w',  struct('modality', 'anat', 'space', '', 'suffix', 'T1w'), ...
       'roi',  struct('modality', 'roi', 'suffix', 'mask'), ...
       'xfm',  struct('modality', 'anat', 'suffix', 'xfm', 'to', 'T1w'));
```
https://bidspm.readthedocs.io/en/latest/dev_doc.html#src.defaults.checkOptions

So in your case you could do:

```matlab
bids_filter_file = struct('fmap', struct('modality', 'fmap'), ...
       'bold', struct('modality', 'func', 'suffix', 'bold', 'desc', 'smoothAROMAnonaggr'), ... % < ---- specify the 'desc' here
       't2w',  struct('modality', 'anat', 'suffix', 'T2w'), ...
       't1w',  struct('modality', 'anat', 'space', '', 'suffix', 'T1w'), ...
       'roi',  struct('modality', 'roi', 'suffix', 'mask'), ...
       'xfm',  struct('modality', 'anat', 'suffix', 'xfm', 'to', 'T1w'));


task_in_the_model ={'food'}

bidspm(bids_dir, output_dir, 'subject', ...
        'action', 'copy', ...
        'participant_label', {subject_label}, ...
        'bids_filter_file', bids_filter_file, ...
        'verbosity', 2, ...
        'space', {'MNI152NLin6Asym'}, ...
        'options', struct([]), ...
        'task', task_in_the_model, ...
        'skip_validation', true)
```

**Reply by Remi-Gau** [reply]
Note that this bids filter file is the same strategy used by fmriprep: https://fmriprep.org/en/stable/faq.html#how-do-i-select-only-certain-files-to-be-input-to-fmriprep

So you can also use a json file to define those filters:

```matlab
>> bids_filter_file = struct('fmap', struct('modality', 'fmap'), ...
       'bold', struct('modality', 'func', 'suffix', 'bold', 'desc', 'smoothAROMAnonaggr'), ...
       't2w',  struct('modality', 'anat', 'suffix', 'T2w'), ...
       't1w',  struct('modality', 'anat', 'space', '', 'suffix', 'T1w'), ...
       'roi',  struct('modality', 'roi', 'suffix', 'mask'), ...
       'xfm',  struct('modality', 'anat', 'suffix', 'xfm', 'to', 'T1w'));
>> bids.util.jsonwrite(fullfile(pwd, 'bids_filter_file.json'), bids_filter_file)
```

This will create this file:

```json
{
  "fmap": {
    "modality": "fmap"
  },
  "bold": {
    "modality": "func",
    "suffix": "bold",
    "desc": "smoothAROMAnonaggr"
  },
  "t2w": {
    "modality": "anat",
    "suffix": "T2w"
  },
  "t1w": {
    "modality": "anat",
    "space": "",
    "suffix": "T1w"
  },
  "roi": {
    "modality": "roi",
    "suffix": "mask"
  },
  "xfm": {
    "modality": "anat",
    "suffix": "xfm",
    "to": "T1w"
  }
}
```

that you should be able to use like this: 

```matlab
task_in_the_model ={'food'}

bidspm(bids_dir, output_dir, 'subject', ...
        'action', 'copy', ...
        'participant_label', {subject_label}, ...
        'bids_filter_file', fullfile(pwd, 'bids_filter_file.json'), ...
        'verbosity', 2, ...
        'space', {'MNI152NLin6Asym'}, ...
        'options', struct([]), ...
        'task', task_in_the_model, ...
        'skip_validation', true)
```

**Comment by Remi-Gau** [comment]
Your bids stats model looks mostly fine:

I removed:
- transformation
- the contrast
and modified the GroupBy at the dataset level to for each contrast from the subject level.

```json
{
  "Name": "default_food_model",
  "BIDSModelVersion": "1.0.0",
  "Description": "default BIDS stats model for food task",
  "Input": {
    "task": ["food"],
    "space": ["MNI152NLin6Asym"]
  },
  "Nodes": [
    {
      "Level": "Run",
      "Name": "run",
      "GroupBy": [
        "run",
        "session",
        "subject"
      ],
      "Model": {
        "X": [
          "trial_type.hi_savory",
          "trial_type.hi_sweet",
          "trial_type.lo_savory",
          "trial_type.lo_sweet",
          "1",
          "trans_?",
          "rot_?",
          "non_steady_state_outlier*",
          "motion_outlier*"
        ],
        "Type": "glm",
        "HRF": {
          "Variables": [
            "trial_type.hi_savory",
            "trial_type.hi_sweet",
            "trial_type.lo_savory",
            "trial_type.lo_sweet"
          ],
          "Model": "spm"
        },
        "Options": {
          "HighPassFilterCutoffHz": 0.008,
          "Mask": {
            "desc": ["brain"],
            "suffix": ["mask"]
          }
        },
        "Software": {
          "SPM": {
            "SerialCorrelation": "FAST",
            "InclusiveMaskingThreshold": 0.8
          }
        }
      },
      "DummyContrasts": {
        "Test": "t",
        "Contrasts": [
          "trial_type.hi_savory",
          "trial_type.hi_sweet",
          "trial_type.lo_savory",
          "trial_type.lo_sweet"
        ]
      }
    },
    {
      "Level": "Subject",
      "Name": "subject",
      "GroupBy": [
        "subject",
        "contrast"
      ],
      "Model": {
        "X": [1],
        "Type": "glm"
      },
      "DummyContrasts": {
        "Test": "t"
      }
    },
    {
      "Level": "Dataset",
      "Name": "dataset",
      "GroupBy": ["contrast"],
      "Model": {
        "X": [1],
        "Type": "glm"
      },
      "DummyContrasts": {
        "Test": "t"
      }
    }
  ],
  "Edges": [
    {
      "Source": "run",
      "Destination": "subject"
    },
    {
      "Source": "subject",
      "Destination": "dataset"
    }
  ]
}
```

**Comment by Rageve** [comment]
When I try to copy using your suggestions I receive the error message

```matlab
15:19:43] bidspm - INFO				getData
WILL WORK ON SUBJECTS
	- 1085


[15:19:43] bidspm - INFO

No data found for this query:
	- desc: {preproc, brain}
	- modality: {anat}
	- space: {MNI152NLin6Asym}
	- suffix: {T1w, bold, mask}
	- sub: {1085}

in dataset:
	/server/fo2-17/data/NASAL_2018/DAGen_Bromo/BIDS/Tuebingen/inputs/fmriprep
```

The data is actually in the dataset folder but with different sessions
.../fmriprep/sub-1085/ses-V0/func
.../fmriprep/sub-1085/ses-V9/func
.../fmriprep/sub-1085/anat

anat is in a superior folder because fmriprep is searching for all  
T1-images in a certain subject and create one anat folder below the  
subject folder.
In the other .../sub-1085/ses-V0/anat folder there is only a file  
sub-1085_ses-V0_from-orig_to-T1w_mode-image_xfm.txt with affine  
transformation parameters.
Should I define the two sessions (and runs) in the bids_filter_file?  
How to handle the different level of the anat folder? I just realized  
that in the anat the data has a different space (MNI152NLin2009cAsym -  
but the default output for anat processing in fmriprep). Defining both  
spaces or the 2009 space in the structure for space does not change  
the error.

**Reply by Remi-Gau** [reply]
Ah this is one case I have not yet encountered yet.

I think this is related to something that still needs to be implemented in bids matlab.

https://github.com/bids-standard/bids-matlab/issues/493

**Reply by Remi-Gau** [reply]
Until I improve bids matlab to handle this
I think that you should be able to specify which modality to copy by creating an option structure.

```matlab
Options = struct('query', struct('modality', {'func'}));
```

You can then pass this structure in your bidspm  command.

**Reply by Remi-Gau** [reply]
```matlab
bidspm(bids_dir, output_dir, 'subject', ...
        'action', 'copy', ...
        'participant_label', {subject_label}, ...
        'bids_filter_file', bids_filter_file, ...
        'verbosity', 2, ...
        'space', {'MNI152NLin6Asym'}, ...
        'options', Options, ... %. <------ see here
        'task', task_in_the_model, ...
        'skip_validation', true)
```

**Reply by Remi-Gau** [reply]
Something I can also implement fairly quickly is for the copy action to throw warnings and not errors when no file is found to copy.

**Comment by Rageve** [comment]
using the options flag, the copying works with the defaults files  
(including the confounds_timeseries.tsv files)

```
  copying  
/server/fo2-17/data/NASAL_2018/DAGen_Bromo/BIDS/Tuebingen/inputs/fmriprep/sub-1085/ses-V0/func/sub-1085_ses-V0_task-food_run-1_space-MNI152NLin2009cAsym_desc-brain_mask.nii.gz -->  
/server/fo2-17/data/NASAL_2018/DAGen_Bromo/BIDS/Tuebingen/outputs/derivatives/bidspm-preproc/derivatives/bidspm-preproc/../bidspm-preproc/sub-1085/ses-V0/func/sub-1085_ses-V0_task-food_run-1_space-MNI152NLin2009cAsym_desc-brain_mask.nii.gz
```

but not with the already smoothed files that were stated in in the  
bids_filter_file

```
bids_filter_file = struct('fmap', struct('modality', 'fmap'), ...
        'bold', struct('modality', 'func', 'suffix', 'bold', 'desc',  
'smoothAROMAnonaggr'), ...
        't2w',  struct('modality', 'anat', 'suffix', 'T2w'), ...
        't1w',  struct('modality', 'anat', 'space',  
'MNI152NLin2009cAsym', 'suffix', 'T1w'), ...
        'roi',  struct('modality', 'roi', 'suffix', 'mask'), ...
        'xfm',  struct('modality', 'anat', 'suffix', 'xfm', 'to', 'T1w'));
```

should I modify the options to select the specific 'bold' files

**Comment by Rageve** [comment]
The bids_filter_file shows the correct field

```
bids_filter_file.bold

ans =

   struct with fields:

     modality: 'func'
        space: 'MNI152NLin6Asym'
       suffix: 'bold'
         desc: 'smoothAROMAnonaggr'
```

but when running a different desc reference is listed

```
No data found for this query:
	- desc: {preproc, brain}
	- modality: {func}
	- space: {MNI152NLin6Asym}
	- suffix: {T1w, bold, mask}
	- sub: {1424}
	- task: {food}
```

**Reply by Remi-Gau** [reply]
OK then this is a bug and I need to fix this.

Will open a separate issue to track this.

**Comment by Remi-Gau** [comment]
OK so this problem should be fixed.

- you should be able to copy `smoothAROMAnonaggr` files by specifying it in the bids_filter_json

I have a test where it works:

https://github.com/cpp-lln-lab/bidspm/blob/e684f7abd6a752ec2de94afddf95e4a802e31299/tests/tests_cli/test_bidspm_copy.m#L60-L71

- you also should not have to specify `opt.query.modality = {'func'}` to avoid crashes when the anat cannot be copied, the copying process has been made tolerant to some errors.

so I am going to close this discussion, but feel free to reopen it or open a new one or an issue if you still encounter problems with this.
